src/support/err.c

```diff
@@ -162,45 +162,124 @@ __wt_event_handler_set(WT_SESSION_IMPL *session, WT_EVENT_HANDLER *handler)
+}
+
+/*
+ * __eventv_gen_json_str --
+ *     Generate a JSON style string.
  */
 static int
-__eventv(WT_SESSION_IMPL *session, bool msg_event, int error, const char *func, int line,
-  const char *fmt, va_list ap) WT_GCC_FUNC_ATTRIBUTE((cold))
+__eventv_gen_json_str(WT_SESSION_IMPL *session, char *buffer, size_t *buffer_len, int error,
+  const char *func, int line, const char *fmt, va_list ap) WT_GCC_FUNC_ATTRIBUTE((cold))
 {
     struct timespec ts;
     WT_DECL_RET;
-    WT_EVENT_HANDLER *handler;
-    WT_SESSION *wt_session;
-    size_t len, remain;
-    char *p, tid[128];
+    size_t msg_len, remain, remain_msg, unpacked_msg_len;
+    u_char *unpacked_json_str;
+    char *p, *p_msg, tid[128], msg_str[2 * 1024];
     const char *err, *prefix;
 
-    /*
-     * We're using a stack buffer because we want error messages no matter
-     * what, and allocating a WT_ITEM, or the memory it needs, might fail.
-     *
-     * !!!
-     * SECURITY:
-     * Buffer placed at the end of the stack in case snprintf overflows.
-     */
-    char s[4 * 1024];
-    p = s;
-    remain = sizeof(s);
+    p = buffer;
+    p_msg = msg_str;
+    unpacked_json_str = NULL;
+
+    remain = *buffer_len;
+    remain_msg = sizeof(msg_str);
 
     /*
-     * !!!
-     * This function MUST handle a NULL WT_SESSION_IMPL handle.
-     *
-     * Without a session, we don't have event handlers or prefixes for the
-     * error message.  Write the error to stderr and call it a day.  (It's
-     * almost impossible for that to happen given how early we allocate the
-     * first session, but if the allocation of the first session fails, for
-     * example, we can end up here without a session.)
+     * We have several attributes for the event handler message: a timestamp and the process and
+     * thread ids, the database error prefix, the data-source's name, and the session's name. Write
+     * them as separates JSON attributes.
      */
-    if (session == NULL)
-        goto err;
+    __wt_epoch(session, &ts);
+    WT_ERR(__wt_thread_str(tid, sizeof(tid)));
+    WT_ERROR_APPEND(p, remain, "{");
+    WT_ERROR_APPEND(p, remain, "\"ts_sec\":%" PRIuMAX ",", (uintmax_t)ts.tv_sec);
+    WT_ERROR_APPEND(p, remain, "\"ts_usec\":%" PRIuMAX ",", (uintmax_t)ts.tv_nsec / WT_THOUSAND);
+    WT_ERROR_APPEND(p, remain, "\"thread\":\"%s\",", tid);
+
+    if ((prefix = S2C(session)->error_prefix) != NULL)
+        WT_ERROR_APPEND(p, remain, "\"session_err_prefix\":\"%s\",", prefix);
+    prefix = session->dhandle == NULL ? NULL : session->dhandle->name;
+    if (prefix != NULL)
+        WT_ERROR_APPEND(p, remain, "\"session_dhandle_name\":\"%s\",", prefix);
+    if ((prefix = session->name) != NULL)
+        WT_ERROR_APPEND(p, remain, "\"session_name\":\"%s\",", prefix);
+
+    if (error != 0) {
+        err = __wt_strerror(session, error, NULL, 0);
+        WT_ERROR_APPEND(p, remain, "\"error_str\":\"%s\",", err);
+        WT_ERROR_APPEND(p, remain, "\"error_code\":\"%d\",", error);
```

Comment:
Realized we should probably drop the quotes around the integer value, better for it to be an JSON integer field than a string field.
` WT_ERROR_APPEND(p, remain, "\"error_code\":%d,", error);`

Reply:
Good point! Fixed in f45f501a9e8eeb62e7177a1b127eedb0203fc534